README.markdown
# Carbon

Carbon is a Ruby API client and command-line console for the [Brighter Planet impact estimate web service](http://impact.brighterplanet.com), which is located at http://impact.brighterplanet.com. By querying the web service, it can estimate the carbon emissions, energy usage, and other environmental impacts of many real-life objects, such as cars and houses, based on particular characteristics that they may have.

Full documentation: [RDoc](http://rdoc.info/projects/brighterplanet/carbon) 

## Quick start 1: experimenting with the console

<b>You'll need a Brighter Planet API key. See the "API keys" section below for details.</b>

First get the gem:

    $ gem install carbon

Then start the console:

    $ carbon
    carbon->

Provide your key:

    carbon-> key '123ABC'
      => Using key 123ABC
      
Start a flight calculation:

    carbon-> flight
      => 1210.66889895298 kg CO2e
    flight*>

Start providing characteristics:

    flight*> origin_airport 'jfk'
      => 1593.46008200024 kg CO2e
    flight*> destination_airport 'lax'
      => 1766.55536727522 kg CO2e

Review what you've entered:

    flight*> characteristics
      => origin_airport: jfk
         destination_airport: lax

See how the calculation's being made:

    flight*> methodology
      => emission: from fuel and passengers with coefficients
         [ ... ]
         cohort: from t100
         
See intermediate calculations:

    flight*> reports
      => emission: 1766.55536727522
         [ ... ]
         cohort: {"members"=>262}

Generate a methodology URL:

    flight*> url
      => http://impact.brighterplanet.com/flights.json?origin_airport=jfk&destination_airport=lax&key=123ABC

And when you're done:

    flight*> done
      => Saved as flight #0
    carbon->

You can recall this flight anytime during this same session:

    carbon-> flight 0
      => 1766.55536727522 kg CO2e
    flight*> characteristics
      => origin_airport: jfk
         destination_airport: lax
         
For more, see the "Console" section below.

## Quick start 2: using the library in your application

<b>You'll need a Brighter Planet API key. See the "API keys" section below for details.</b>

Carbon works by extending any Ruby class to be an emission source. You `include Carbon` and then use the `emit_as` DSL...

    # see Carbon::ClassMethods#emit_as for more details
    class MyFlight
      def airline
        # ... => MyAirline(:name, :icao_code, ...)
      end
      def aircraft
        # ... => MyAircraft(:name, :icao_code, ...)
      end
      def origin
        # ... => String
      end
      def destination
        # ... => String
      end
      def segments_per_trip
        # ... => Integer
      end
      def trips
        # ... => Integer
      end
      include Carbon
      emit_as 'Flight' do
        provide :segments_per_trip
        provide :trips
        provide :origin, :as => :origin_airport, :key => :iata_code
        provide :destination, :as => :destination_airport, :key => :iata_code
        provide(:airline, :key => :iata_code) { |f| f.airline.try(:iata_code) }
        provide(:aircraft, :key => :icao_code) { { |f| f.aircraft.try(:icao_code) }
      end
    end

See [RDoc on `Carbon::ClassMethods#emit_as`](http://rdoc.info/github/brighterplanet/carbon/Carbon/ClassMethods#emit_as-instance_method) for all the details.

The final URL will be something like

    http://impact.brighterplanet.com/flights.json?segments_per_trip=1&trips=1&origin_airport[iata_code]=MSN&destination_airport[iata_code]=ORD&airline[iata_code]=UA&aircraft[icao_code]=B737

When you want to calculate impacts, simply call `MyFlight#impact`.

    ?> my_flight = MyFlight.new([...])
    => #<MyFlight [...]>
    ?> my_impact = my_flight.impact(:timeframe => Timeframe.new(:year => 2009))
    => #<Hashie::Mash [...]>
    ?> my_impact.decisions.carbon.object.value
    => 1014.92
    ?> my_impact.decisions.carbon.object.units
    => "kilograms"
    ?> my_impact.methodology
    => "http://impact.brighterplanet.com/flights?[...]"

See [RDoc on `Carbon#impact`](http://rdoc.info/github/brighterplanet/carbon/Carbon#impact-instance_method) for all the details.

## API keys

You should get an API key from http://keys.brighterplanet.com and set it globally:

    Carbon.key = '12903019230128310293'

Now all of your queries will use that key.

## Gotcha: timeframes and 0.0kg results

You submit this query about a flight in 2009, but the result is 0.0 kilograms. Why?

  $ carbon 
  carbon-> flight
  [...]
  flight*> date '2009-05-03'
    => 0.0 kg CO2e
  flight*> url
    => http://impact.brighterplanet.com/flights?date=2009-05-03

It's telling you that a flight in 2009 did not result in any 2011 emissions (the default timeframe is the current year).

  flight*> timeframe '2009'
    => 847.542137647608 kg CO2e
  flight*> url
    => http://impact.brighterplanet.com/flights?date=2009-05-03&timeframe=2009-01-01/2010-01-01

So, 850 kilograms emitted in 2009.

## Console

This library includes a special console for performing calculations interactively. Quick Start #1 provides an example session. Here is a command reference:

### Shell mode

`help`
:  Displays a list of emitter types.

`key` _yourkey_
:  Set the [developer key](http://keys.brighterplanet.com) that should be used for this session. Alternatively, put this key in `~/.brighter_planet` and it will be auto-selected on console startup.

_emitter_
:  (e.g. `flight`) Enters emitter mode using this emitter type.

_emitter num_
:  (e.g. `flight 0`) Recalls a previous emitter from this session.

`exit`
:  Quits.
  
### Emitter mode

In Emitter mode, the prompt displays the emitter type in use. If a timeframe has been set, the timeframe is also included in the prompt.

`help`
:  Displays a list of characteristics for this emitter type.

_characteristic value_
:  (e.g. `origin_airport 'lax'`)
:  Provide a characteristic. Remember, this is Ruby we're dealing with, so strings must be quoted.

`timeframe`
:  Display the current timeframe in effect on the emission estimate.

`timeframe` _timeframe_
:  (e.g. `timeframe '2009-01-01/2010-01-01'` or just `timeframe '2009'`) Set a timeframe on the emission estimate.

`emission`
:  Displays the current emission in kilograms CO2e for this emitter.

`lbs`, `pounds`, or `tons`
:  Display the emission using different units.

`characteristics`
:  Lists the characteristics you have provided so far.

`methodology`
:  Summarizes how the calculation is being made.

`reports`
:  Displays intermediate calculations that were made in pursuit of the emission estimate.

`url`
:  Generates a methodology URL suitable for pasting into your browser for further inspection.

`done`
:  Saves this emitter and returns to shell mode.

## Copyright

Copyright (c) 2012 Brighter Planet.


CHANGELOG
3.0.1 / 2012-06-20

* Enhancements

  * Provide "quick decisions" shortcuts
    - result.decisions.carbon.object.value -> result.carbon
    - result.decisions.energy.object.value -> result.energy

3.0.0 / 2012-06-08

* Breaking changes

  * Ruby 1.9 only because it uses celluloid / fibers. Use ~2 if you are on Ruby 1.8

* Bug fixes

  * Make sure #as_impact_query includes API key if it's been set globally

* Enhancements

  * When doing multiple queries, use Celluloid [worker] pools instead of creating new threads and using a homegrown futures class
	* Use RSpec as the API testing framework

2.2.3 / 2012-04-12

* Enhancements

  * Raise an ArgumentError if you make the common mistake of calling Carbon.query(array) {} instead of Carbon.query(array).each {}

* Bug fixes

  * Fix how shell gets characteristics for an emitter

2.2.2 / 2012-04-05

* Enhancements

  * Allow choosing domain (API endpoint)
  * Slightly improved method signature explanations: query_array -> plain_query, o -> obj, os -> array

* Bug fixes

  * Actually send keys set with Carbon.key

2.2.1 / 2012-03-21

* Enhancements

  * Switched from EventMachine to pure Ruby threads, a small speed sacrifice that gains us compatibilty with MRI 1.8, MRI 1.9, and JRuby. Also works better in multithreaded environments - no accidentally turning off somebody's EventMachine reactor :)

2.2.0 / 2012-03-20

* Enhancements

  * Carbon.query(os) now returns a Hash keyed by the "os", instead of just a carefully ordered Array.

* Breaking changes

  * Carbon.query(os) has a different return value!

2.1.0 / 2012-03-14

* Enhancements

  * Simplified API down to just Carbon.query. It has three method signatures.
  * Simplified EventMachine reactor
  * Limited concurrency to 16 connections at a time

* Breaking changes

  * Simplified API down to just Carbon.query :)

2.0.3 / 2012-03-13

* Enhancements

  * Automatically skip redundant queries and cache things for an hour (in-process, unless you set up CacheMethod.config.storage = $memcached or something)

2.0.2 / 2012-03-13

* Bug fixes

  * Don't hang up on Carbon.multi if 0 queries are passed

2.0.1 / 2012-03-12

* Bug fixes

  * Don't send blank (incl. nil) characteristics

* Documentation changes

  * Recommend people use Object#try when passing a block to #provide

2.0.0 / 2012-03-09

* Breaking changes

  * #emission_estimate has been removed in favor of #impact
  * Response structure now mirrors what you get from http://impact.brighterplanet.com

* Enhancements

  * Carbon.multi method for parallelizing requests
  * Tested with MRI 1.8 and MRI 1.9


lib/carbon.rb
require 'active_support/core_ext'

require 'carbon/query'
require 'carbon/query_pool'
require 'carbon/registry'

module Carbon
  DOMAIN = 'http://impact.brighterplanet.com'.freeze
  CONCURRENCY = 16

  # @private
  # Make sure there are no warnings about class vars.
  @@key = nil unless defined?(@@key)

  # @private
  # Make sure there are no warnings about class vars.
  @@domain = nil unless defined?(@@domain)

  # Set the Brighter Planet API key that you can get from http://keys.brighterplanet.com
  #
  # @param [String] key The alphanumeric key.
  #
  # @return [nil]
  def Carbon.key=(key)
    @@key = key
    nil
  end

  # Get the key you've set.
  #
  # @return [String] The key you set.
  def Carbon.key
    @@key
  end

  # Set an alternate API endpoint. You probably shouldn't do this.
  #
  # @param [String] domain ("http://impact.brighterplanet.com") The API endpoint
  #
  # @return [nil]
  def Carbon.domain=(domain)
    @@domain = domain
    nil
  end

  # Where we send queries.
  #
  # @return [String] The API endpoint.
  def Carbon.domain
    @@domain || DOMAIN
  end

  # Get impact estimates from Brighter Planet CM1; low-level method that does _not_ require you to define {Carbon::ClassMethods#emit_as} blocks; just pass emitter/param or objects that respond to +#as_impact_query+.
  #
  # Return values are {http://rdoc.info/github/intridea/hashie/Hashie/Mash Hashie::Mash} objects because they are a simple way to access a deeply nested response.
  #
  # Here's a map of what's included in a response:
  #
  #     certification
  #     characteristics.{}.description
  #     characteristics.{}.object
  #     compliance.[]
  #     decisions.{}.description
  #     decisions.{}.methodology
  #     decisions.{}.object
  #     emitter
  #     equivalents.{}
  #     errors.[]
  #     methodology
  #     scope
  #     timeframe.endDate
  #     timeframe.startDate
  #
  # @overload query(emitter, params)
  #   The simplest form.
  #   @param [String] emitter The {http://impact.brighterplanet.com/emitters.json emitter name}.
  #   @param [optional, Hash] params Characteristics like airline/airport/etc., your API key (if you didn't set it globally), timeframe, compliance, etc.
  #   @option params [Timeframe] :timeframe (Timeframe.this_year) What time period to focus the calculation on. See {https://github.com/rossmeissl/timeframe timeframe} documentation.
  #   @option params [Array<Symbol>] :comply ([]) What {http://impact.brighterplanet.com/protocols.json calculation protocols} to require.
  #   @option params [String, Numeric] <i>characteristic</i> Pieces of data about an emitter. The {http://impact.brighterplanet.com/flights/options Flight characteristics API} lists valid keys like +:aircraft+, +:origin_airport+, etc.
  #   @return [Hashie::Mash] The API response, contained in an easy-to-use +Hashie::Mash+
  #
  # @overload query(obj)
  #   Pass in a single query-able object.
  #   @param [#as_impact_query] obj An object that responds to +#as_impact_query+, generally because you've declared {Carbon::ClassMethods#emit_as} on its parent class.
  #   @return [Hashie::Mash] The API response, contained in an easy-to-use +Hashie::Mash+
  #
  # @overload query(array)
  #   Get impact estimates for multiple query-able objects concurrently.
  #   @param [Array<Array, #as_impact_query>] array An array of plain queries and/or objects that respond to +#as_impact_query+.
  #   @return [Hash{Object => Hashie::Mash}] A +Hash+ of +Hashie::Mash+ objects, keyed on the original query object.
  #
  # @note We make up to 16 requests concurrently (hardcoded, per the Brighter Planet Terms of Service) and it can be more than 90% faster than running queries serially!
  #
  # @note +[emitter, params]+ is called a "plain query."
  #
  # @raise [ArgumentError] If your arguments don't match any of the method signatures.
  #
  # @raise [ArgumentError] If you try to pass a block - you probably want +Carbon.query(array).each {}+ or something.
  #
  # @example A flight taken in 2009
  #   Carbon.query('Flight', :origin_airport => 'MSN', :destination_airport => 'ORD', :date => '2009-01-01', :timeframe => Timeframe.new(:year => 2009), :comply => [:tcr])
  #
  # @example How do I use a +Hashie::Mash+?
  #   1.8.7 :001 > require 'rubygems'
  #    => true 
  #   1.8.7 :002 > require 'hashie/mash'
  #    => true 
  #   1.8.7 :003 > mash = Hashie::Mash.new(:hello => 'world')
  #    => #<Hashie::Mash hello="world"> 
  #   1.8.7 :004 > mash.hello
  #    => "world" 
  #   1.8.7 :005 > mash['hello']
  #    => "world" 
  #   1.8.7 :006 > mash[:hello]
  #    => "world" 
  #   1.8.7 :007 > mash.keys
  #    => ["hello"] 
  #
  # @example Other examples of what's in the response
  #   my_impact.carbon.object.value
  #   my_impact.characteristics.airline.description
  #   my_impact.equivalents.lightbulbs_for_a_week
  #
  # @example Flights and cars (concurrently, as arrays)
  #   queries = [
  #     ['Flight', {:origin_airport => 'MSN', :destination_airport => 'ORD', :date => '2009-01-01', :timeframe => Timeframe.new(:year => 2009), :comply => [:tcr]}],
  #     ['Flight', {:origin_airport => 'SFO', :destination_airport => 'LAX', :date => '2011-09-29', :timeframe => Timeframe.new(:year => 2011), :comply => [:iso]}],
  #     ['Automobile', {:make => 'Nissan', :model => 'Altima', :timeframe => Timeframe.new(:year => 2008), :comply => [:tcr]}]
  #   ]
  #   Carbon.query(queries)
  #
  # @example Flights and cars (concurrently, as query-able objects)
  #   Carbon.query(MyFlight.all+MyCar.all).each do |car_or_flight, impact|
  #     puts "Carbon emitter by #{car_or_flight} was #{impact.decisions.carbon.object.value.round(1)}"
  #   end
  def Carbon.query(*args)
    raise ::ArgumentError, "Don't pass a block directly - instead use Carbon.query(array).each (for example)." if block_given?
    Query.perform(*args)
  end

  # Called when you +include Carbon+ and adds the class method +emit_as+.
  # @private
  def Carbon.included(klass)
    klass.extend ClassMethods
  end

  # Mixed into any class that includes +Carbon+.
  module ClassMethods
    # DSL for declaring how to represent this class an an emitter.
    #
    # See also {Carbon::Registry::Registrar#provide}.
    #
    # You get this when you +include Carbon+ in a class.
    #
    # @param [String] emitter The {http://impact.brighterplanet.com/emitters.json camelcased emitter name}.
    #
    # @return [nil]
    #
    # @example MyFlight
    #   # A a flight in your data warehouse
    #   class MyFlight
    #     def airline
    #       # ... => MyAirline(:name, :icao_code, ...)
    #     end
    #     def aircraft
    #       # ... => MyAircraft(:name, :icao_code, ...)
    #     end
    #     def origin
    #       # ... => String
    #     end
    #     def destination
    #       # ... => String
    #     end
    #     def segments_per_trip
    #       # ... => Integer
    #     end
    #     def trips
    #       # ... => Integer
    #     end
    #     include Carbon
    #     emit_as 'Flight' do
    #       provide :segments_per_trip
    #       provide :trips
    #       provide :origin, :as => :origin_airport, :key => :iata_code
    #       provide :destination, :as => :destination_airport, :key => :iata_code
    #       provide(:airline, :key => :iata_code) { |f| f.airline.try(:iata_code) }
    #       provide(:aircraft, :key => :icao_code) { { |f| f.aircraft.try(:icao_code) }
    #     end
    #   end
    def emit_as(emitter, &blk)
      emitter = emitter.to_s.singularize.camelcase
      registrar = Registry::Registrar.new self, emitter
      registrar.instance_eval(&blk)
    end
  end

  # A query like what you could pass into +Carbon.query+.
  #
  # @param [Hash] extra_params Anything you want to override.
  #
  # @option extra_params [Timeframe] :timeframe
  # @option extra_params [Array<Symbol>] :comply
  # @option extra_params [String] :key In case you didn't define it globally, or want to use a different one here.
  # @option extra_params [String, Numeric] <i>characteristic</i> Override pieces of data about an emitter.
  #
  # @return [Array] Something you could pass into +Carbon.query+.
  def as_impact_query(extra_params = {})
    registration = Registry.instance[self.class.name]
    params = registration.characteristics.inject({}) do |memo, (method_id, translation_options)|
      k = translation_options.has_key?(:as) ? translation_options[:as] : method_id
      if translation_options.has_key?(:key)
        k = "#{k}[#{translation_options[:key]}]"
      end
      v = if translation_options.has_key?(:blk)
        translation_options[:blk].call self
      else
        send method_id
      end
      if v.present?
        memo[k] = v
      end
      memo
    end
    params.merge! extra_params
    if Carbon.key and not params.has_key?(:key)
      params[:key] = Carbon.key
    end
    [ registration.emitter, params ]
  end

  # Get an impact estimate from Brighter Planet CM1; high-level convenience method that requires a {Carbon::ClassMethods#emit_as} block.
  #
  # You get this when you +include Carbon+ in a class.
  #
  # See {Carbon.query} for an explanation of the return value, a +Hashie::Mash+.
  #
  # @param [Hash] extra_params Anything you want to override.
  #
  # @option extra_params [Timeframe] :timeframe
  # @option extra_params [Array<Symbol>] :comply
  # @option extra_params [String] :key In case you didn't define it globally, or want to use a different one here.
  # @option extra_params [String, Numeric] <i>characteristic</i> Override pieces of data about an emitter.
  #
  # @return [Hashie::Mash]
  #
  # @example Getting impact estimate for MyFlight
  #   ?> my_flight = MyFlight.new([...])
  #   => #<MyFlight [...]>
  #   ?> my_impact = my_flight.impact(:timeframe => Timeframe.new(:year => 2009))
  #   => #<Hashie::Mash [...]>
  #   ?> my_impact.decisions.carbon.object.value
  #   => 1014.92
  #   ?> my_impact.decisions.carbon.object.units
  #   => "kilograms"
  #   ?> my_impact.methodology
  #   => "http://impact.brighterplanet.com/flights?[...]"
  def impact(extra_params = {})
    plain_query = as_impact_query extra_params
    Carbon.query(*plain_query)
  end
end
